### ISSUE TITLE
Integrate password checkup into Brave

### ISSUE THREAD

**Issue by diracdeltas** [brave/brave-browser#12001]
chrome://settings/security has a feature (on by default IIRC) that warns if your credentials have been exposed in a breach:

<img width="646" alt="Screen Shot 2020-10-05 at 10 50 00 AM" src="https://user-images.githubusercontent.com/549654/95114279-89e3dd00-06f8-11eb-94c3-79728b6da3fd.png">

we should look into enabling this in Brave

**Comment by diracdeltas** [comment]
here is one option that doesn't require any calls to google:

1. host the breached passwords database from https://haveibeenpwned.com/Passwords on https://passwords.brave.com (or a similar domain owned by brave)
2. implement the partial-hash search feature from https://haveibeenpwned.com/API/v3#PwnedPasswords on passwords.brave.com
3. when a user enables the Password Checkup feature, Brave downloads the hash prefix list
4. when a user logs in, their password is hashed, and the first 5 characters are checked against the local hash prefix list.
5. if there is a match in the local hash prefix list, send a request to passwords.brave.com for the full list of hashes which matches the hash prefix
5. check the full password hash against the list of full hashes matching the prefix
6. if there is a match, warn them

this is conceptually very similar to safebrowsing

**Comment by fmarier** [comment]
I agree that we should implement that in Brave. The Chrome feature requires connecting the browser to a Google Account.

> 5. if there is a match in the local hash prefix list, send a request to passwords.brave.com for the full list of hashes which matches the hash prefix

That step could use `pcdn.brave.com` for extra IP-address privacy.

**Comment by fmarier** [comment]
Here's the code we'd likely have to replace in Chromium to switch to a different password check service: https://source.chromium.org/chromium/chromium/src/+/master:components/password_manager/core/browser/leak_detection/;bpv=1;bpt=0

**Comment by fmarier** [comment]
Edge also uses homomorphic encryption to talk to the equivalent Microsoft service: https://www.microsoft.com/en-us/research/blog/password-monitor-safeguarding-passwords-in-microsoft-edge/

**Comment by fmarier** [comment]
Chrome now also makes it easy for users to automatically update compromised passwords: https://blog.google/products/chrome/automated-password-changes

**Comment by DaCon** [comment]
Here's another approach to the issue that should be a browser feature running in the background whenever a username or password is used on a website: https://github.com/ibnaleem/gosearch
